src/wheel/wheelfile.py

```diff
@@ -1,169 +1,311 @@
-                        except ValueError:
-                            raise WheelError('Unsupported hash algorithm: {}'.format(algorithm))
-
-                        if algorithm.lower() in {'md5', 'sha1'}:
-                            raise WheelError(
-                                'Weak hash algorithm ({}) is not permitted by PEP 427'
-                                .format(algorithm))
-
-                        self._file_hashes[path] = (
-                            algorithm, urlsafe_b64decode(hash_sum.encode('ascii')))
-
-    def open(self, name_or_info, mode="r", pwd=None):
-        def _update_crc(newdata, eof=None):
-            if eof is None:
-                eof = ef._eof
-                update_crc_orig(newdata)
-            else:  # Python 2
-                update_crc_orig(newdata, eof)
-
-            running_hash.update(newdata)
-            if eof and running_hash.digest() != expected_hash:
-                raise WheelError("Hash mismatch for file '{}'".format(native(ef_name)))
-
-        ef_name = as_unicode(name_or_info.filename if isinstance(name_or_info, ZipInfo)
-                             else name_or_info)
-        if mode == 'r' and not ef_name.endswith('/') and ef_name not in self._file_hashes:
-            raise WheelError("No hash found for file '{}'".format(native(ef_name)))
-
-        ef = ZipFile.open(self, name_or_info, mode, pwd)
-        if mode == 'r' and not ef_name.endswith('/'):
-            algorithm, expected_hash = self._file_hashes[ef_name]
-            if expected_hash is not None:
-                # Monkey patch the _update_crc method to also check for the hash from RECORD
-                running_hash = hashlib.new(algorithm)
-                update_crc_orig, ef._update_crc = ef._update_crc, _update_crc
-
-        return ef
-
-    def write_files(self, base_dir):
-        logger.info("creating '%s' and adding '%s' to it", self.filename, base_dir)
+                self._zip.close()
+            finally:
+                self._record_entries.clear()
+
+    def __enter__(self) -> 'WheelFile':
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        self.close()
+
+    @staticmethod
+    def _get_zipinfo_datetime(timestamp: float) -> Tuple[int, int, int, int, int, int]:
+        # Some applications need reproducible .whl files, but they can't do this without forcing
+        # the timestamp of the individual ZipInfo objects. See issue #143.
+        timestamp = int(os.environ.get('SOURCE_DATE_EPOCH', timestamp))
+        return time.gmtime(timestamp)[0:6]
+
+    def write_file(self, archive_name: str, contents: Union[bytes, str, PathLike]) -> None:
+        timestamp = time.time()
+        if isinstance(contents, bytes):
+            data = contents
+        elif isinstance(contents, str):
+            data = contents.encode('utf-8')
+        elif isinstance(contents, PathLike):
+            path = Path(contents)
+            timestamp = path.stat().st_mtime
```

Comment:
If timestamp is an argument to the function, maybe we should only use the file timestamp if the argument was left as default?

Reply:
Now that I've removed the ability to pass `PathLike` objects as contents, we can't do that anymore.